Openthread: Fix MTD build

ramonhpr opened the pull request:
This pull request is related to this issue #8833

ramonhpr committed:
net: openThread: Fix MTD build

This patch fix the problem when the user select to use the MTD
in OpenThread configuration

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

KaSroka reviewed (approved):
Overall looks good, there are some CI errors

rlubos reviewed (changes requested):
Thank you for submitting this!

I've left some comments inline, mostly related to conding style incompablilities issues, to help you pass the CI.

I have some remarks about commit titles though. I know there hasn't been a common approach to mark OpenThread related commits, but I think we can work on having one common prefix for that. In general we prefix commits to networking subsystem with `net: `, so I think that `net: openthread: ` should be a reasonable choice. So please replace the `OpenThread: ` prefix with `net: openthread: ` in message titles.

Also, in commit bumping OT version, I would include the actual commit hash we are bumping to instead of generic most recent hash. So please change that commit title to something like `net: openthread: Bump OpenThread to commit 19c73f8b`.

rlubos commented on subsys/net/l2/openthread/openthread_utils.c:
```diff
@@ -19,6 +19,56 @@
+        {
+            byte |= 10 + (*aHex - 'A');
+        }
+        else if ('a' <= *aHex && *aHex <= 'f')
```
4. `else` should be placed in the same line as closing brace, so we are looking for something like `} else if (..) {`

jukkar commented on subsys/net/l2/openthread/openthread_utils.c:
```diff
@@ -19,6 +19,56 @@
 
 #include "openthread_utils.h"
 
+int hex2bin(const char *aHex, uint8_t *aBin, uint16_t aBinLength)
```
We have very similar function already in include/net/net_ip.h called net_bytes_from_str(). Could you make it work with plain hex strings too (it now parses strings in "aa:bb:cc:dd" format) so that we would not need to introduce similar kind of function.

ramonhpr commented:
Thanks for the reviews. I'm working on them.
As I could see the configuration of board frdm_kw41z is being build by LOG_LEVEL_INFO but in the openthread build I saw that there is defined that LOG_LEVEL must be greater than LOG_LEVEL_INFO as described here: https://github.com/openthread/openthread/blob/e69019f962c71c3cad8ee5855f18b00bb483a700/src/core/thread/mle.hpp#L1511
What seems different is that the logging from openthread is different from zephyr defined log. I think that CI error will keep blocking this patch if that won't be fixed.
I was thinking in define the LOG_LEVEL_NOTE to zephyr stay with same enumeration from openthread logging:
https://github.com/openthread/openthread/blob/92a70fb0ad8dba25082e4fbc726997b7434a483d/include/openthread/platform/logging.h#L64

Is there any docs file to update about it in code? Because I didn't find any.

rlubos commented:
I see the problem. I'm not sure if I understood correctly, but adding an extra log level to Zephyr just beacause some external library  (OpenThread) defined it doesn't sound like a good idea and I'm afraid it wouldn't be well received. What if some other library provided different enumeration?

But the fact is that OpenThread complicated things slightly for us with this log level change. That's what I would do (I've briefly tested it on your branch and it seems to sove the problem):
1. Add new log level and appropriate numeric values to OpenThread's Kconfig under `config	OPENTHREAD_LOG_LEVEL`. We're defining OpenThread's logging level there, not Zephyr's.
2. In `openthread/platform/logging.c`, inside `otPlatLog`, handle this new log level, but call `SYS_LOG_INF` instead (so it would be called for both, OpenThread's NOTE and INFO level), so we get something like:
```
	case OT_LOG_LEVEL_WARN:
		SYS_LOG_WRN("%s", logString);
		break;
	case OT_LOG_LEVEL_NOTE:
		SYS_LOG_INF("%s", logString);
		break;
	case OT_LOG_LEVEL_INFO:
		SYS_LOG_INF("%s", logString);
		break;
```
3. For some reason, there is a mapping betwen OpenThread's log level (CONFIG_OPENTHREAD_LOG_LEVEL) to SYS_LOG_LEVEL on top of `openthread/platfrom/radio.c`. That's not correct, and will most likely cause problems if left like that. 
I'd set it to CONFIG_OPENTHREAD_L2_LOG_LEVEL - that's separate config used for logging OpenThread L2 integration - I guess we could use it.

ramonhpr committed:
net: openthread: Add NETWORKNAME and XPANID config

This patch add the option to the user sets the network name and
the extended PAN ID

Signed-off-by: Ramon Ribeiro <rhpr@cesar.org.br>

ramonhpr commented:
Yes, that's what I mean when you mentioned in 1 but I don't realize that 2 and 3 topics @rlubos , thanks for that, I'm new here in zephyr.
I am using the function `net_bytes_from_str`, recommended by @jukkar,  to avoid the use of `hex2bin` that I got from openthread code.
I'm testing the echo_client sample in nrf52840 as a MTD and send to a server board as FTD, and when I flash the two boards everything works fine.
But I'm getting the below problem when I reset my client board. So I enabled the debug mode and I saw that it is not adding the IPv6 that I defined in CONFIG_NET_APP_MY_IPV6_ADDR and my server can't send back the packets to my server board.
The problem just happen when I reset the client board, when I flash in the first time the board connect with my other server board FTD (full thread device).
Is there some way to I debug this using the thread ID? 

```shell
[otPlat/radio] [DBG] otPlatRadioGetPromiscuous: PromiscuousMode=0
[net] [INF] ot_state_changed_handler: State changed! Flags: 0x015f133f Current role: 1
[net] [DBG] ot_state_changed_handler: (0x20000850): Ipv6 address removed
[net] [DBG] ot_state_changed_handler: (0x20000850): Ipv6 address added
[net] [DBG] add_ipv6_maddr_to_zephyr: (0x20000850): Adding multicast ff33:40:fdde:ad00:beef::1
[net] [DBG] add_ipv6_maddr_to_zephyr: (0x20000850): Adding multicast ff32:40:fdde:ad00:beef::1
[net] [DBG] add_ipv6_maddr_to_zephyr: (0x20000850): Adding multicast ff02::1
[net] [DBG] add_ipv6_maddr_to_zephyr: (0x20000850): Adding multicast ff03::1
[net] [DBG] add_ipv6_maddr_to_zephyr: (0x20000850): Adding multicast ff03::fc
***** BUS FAULT *****
  Executing thread ID (thread): 0x20000fbc
  Faulting instruction address:  0x399a
  Precise data bus error
  Address: 0x2100006b
Fatal fault in thread 0x20000fbc! Aborting.
***** Booting Zephyr OS v1.12.0-673-g00e168e *****
Zephyr Shell, Zephyr version: 1.12.99
```

rlubos commented:
> Is there some way to I debug this using the thread ID?

Well, what I do in such cases is a static analysis of `zephyr.lst` and `zephyr.map` files produced during build to identify which thread caused abort and what instruction cauesed it (using the data provided by fault handler). Then you can use `gdb` and try to figure out why the instruction is failing.

ramonhpr commented:
I saw this files and they showed that the problem seems to be in net_mgmt thread_data but and for some reason the callback handler is not been called when the client establish a connect to the server after I bumped to that hash from openthread.

I've made some tests and there is no need to bump to a recent openthread hash for MTD works, maybe some missing configuration in my server wasn't making client connects in the first time I run.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -284,6 +284,9 @@ enum net_verdict ieee802154_radio_handle_ack(struct net_if *iface,
 	struct openthread_context *ot_context = net_if_l2_data(iface);
+	u8_t xpanid[8];
+
+	net_bytes_from_str(xpanid, 8, (char *) CONFIG_OPENTHREAD_XPANID);
```
No space needed after type cast.

rlubos reviewed (approved):
Looks good, one minor style issue spotted.
And I'd appreciate if you could change commit prefix to `net: openthread:` in the second commit as well.

jukkar commented:
recheck
